nexus/db-queries/src/db/pool_connection.rs

```diff
@@ -224,90 +39,152 @@ impl CustomizeConnection<Connection<DbConnection>, ConnectionError>
+                        .await
+                        .expect("Failed to BEGIN transaction");
+                    client
+                        .batch_execute(step.sql())
+                        .await
+                        .expect("Failed to execute step");
+                    client
+                        .batch_execute("COMMIT;")
+                        .await
+                        .expect("Failed to COMMIT transaction");
                 }
-                panic!("Missing 'name' in row: {row:?}");
-            })
-            .collect::<Vec<_>>();
-        observed_public_enums.sort();
 
-        let mut expected_enums: Vec<String> =
-            CUSTOM_TYPE_KEYS.into_iter().map(|s| s.to_string()).collect();
-        expected_enums.sort();
+                // Normally, Nexus actually bumps the version number.
+                //
+                // We do so explicitly here.
+                let sql = format!(
+                    "UPDATE omicron.public.db_metadata \
+                     SET version = '{}', target_version = NULL \
+                     WHERE singleton = true;",
+                    version
+                );
+                client
+                    .batch_execute(&sql)
+                    .await
+                    .expect("Failed to bump version number");
+            }
+        }
 
-        pretty_assertions::assert_eq!(
-            observed_public_enums,
-            expected_enums,
-            "Enums did not match.\n\
-            If the type is present on the left, but not the right:\n\
-            \tThe enum is in the DB, but not in CUSTOM_TYPE_KEYS.\n\
-            \tConsider adding it, so we can pre-populate the OID cache.\n\
-            If the type is present on the right, but not the left:\n\
-            \tThe enum is not the DB, but it is in CUSTOM_TYPE_KEYS.\n\
-            \tConsider removing it, because the type no longer exists"
-        );
+        // Wait for the datastore, which should shortly realize the version is
+        // correct.
+        let datastore =
+            tokio::time::timeout(Duration::from_secs(30), datastore_task)
+                .await
+                .expect("timed out waiting for datastore")
+                .expect("datastore creation panicked");
+
+        // Attempt to insert a sled resource. This involves the
+        // `sled_resource_kind` enum, whose OID was changed by the schema
+        // migration in version 53.0.0. If the diesel OID cache was populated
+        // when we the datastore started polling for the version, this would
+        // fail with a `type with ID $NUM does not exist` error.
+        //
+        // This test isn't perfect: (a) it's possible on an overloaded system
+        // the datastore doesn't start polling CRDB until after we've finished
+        // applying updates (although that's slow, so this seems very unlikely),
+        // and (b) it's possible some _other_ OID has been cached that we're not
```

Comment:
(a) seems like a real problem that will lead to a flaky test.

I'm not sure how much plumbing you'd need to do, but what if you create the pool first here, check out a connection (so you know you've got one), then use that pool to create a new datastore?  Then you don't need a background task and there's no race.